odo describe binding crash

/kind bug

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the Google group if you have a question rather than a bug or feature request.

The group is at: https://groups.google.com/forum/#!forum/odo-users

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?

**Operating System:** Win11Pro

**Output of `odo version`:**

`odo v3.4.0 (b8662ef74)`

## How did you run odo exactly?

`odo describe binding`


## Actual behavior

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x18 pc=0x187d19a]

goroutine 1 [running]:
github.com/redhat-developer/odo/pkg/binding.(*BindingClient).GetBindingsFromDevfile(0xc000823e80, {{{0xc00014e528, 0x5}, {0xc00046f3e0, 0x2d}, {0xc00046f380, 0x2d}, {0xc00029ca00, 0x1386, 0x1500}, ...}, ...}, ...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/binding/binding.go:148 +0x3ba
github.com/redhat-developer/odo/pkg/odo/cli/describe.(*BindingOptions).runWithoutName(0xc000937bc0, {0x22924d8, 0xc0008d33e0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/cli/describe/binding.go:99 +0x105
github.com/redhat-developer/odo/pkg/odo/cli/describe.(*BindingOptions).Run(0x22924d8?, {0x22924d8?, 0xc0008d33e0?})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/cli/describe/binding.go:69 +0x47
github.com/redhat-developer/odo/pkg/odo/genericclioptions.GenericRun({0x2292ee8, 0xc000937bc0}, 0xc000933400, {0x34a0428, 0x0, 0x0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/genericclioptions/runnable.go:278 +0x15cd
github.com/redhat-developer/odo/pkg/odo/cli/describe.NewCmdBinding.func1(0xc000933400?, {0x34a0428?, 0x0?, 0x0?})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/cli/describe/binding.go:117 +0x38
github.com/spf13/cobra.(*Command).execute(0xc000933400, {0x34a0428, 0x0, 0x0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:872 +0x694
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001bcf00)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:990 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:918
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:911
main.main()
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/cmd/odo/odo.go:83 +0x579
```


## Expected behavior

Binding should be listed

## Any logs, error output, etc?

[devfile.zip](https://github.com/redhat-developer/odo/files/10246319/devfile.zip)

--- a/pkg/alizer/mock.go
+++ b/pkg/alizer/mock.go
@@ -36,7 +36,7 @@ func (m *MockClient) EXPECT() *MockClientMockRecorder {
 	return m.recorder
 }
 
-// DetectFramework mocks base method
+// DetectFramework mocks base method.
 func (m *MockClient) DetectFramework(ctx context.Context, path string) (model.DevFileType, string, api.Registry, error) {
 	m.ctrl.T.Helper()
 	ret := m.ctrl.Call(m, "DetectFramework", ctx, path)

--- a/pkg/api/binding.go
+++ b/pkg/api/binding.go
@@ -1,9 +1,5 @@
 package api
 
-import (
-	corev1 "k8s.io/api/core/v1"
-)
-
 // ServiceBinding describes a service binding, from group binding.operators.coreos.com/v1alpha1 or servicebinding.io/v1alpha3
 type ServiceBinding struct {
 	Name   string                `json:"name"`
@@ -12,11 +8,19 @@ type ServiceBinding struct {
 }
 
 type ServiceBindingSpec struct {
-	Application            corev1.ObjectReference   `json:"application"`
-	Services               []corev1.ObjectReference `json:"services"`
-	DetectBindingResources bool                     `json:"detectBindingResources"`
-	BindAsFiles            bool                     `json:"bindAsFiles"`
-	NamingStrategy         string                   `json:"namingStrategy,omitempty"`
+	Application            ServiceBindingReference   `json:"application"`
+	Services               []ServiceBindingReference `json:"services"`
+	DetectBindingResources bool                      `json:"detectBindingResources"`
+	BindAsFiles            bool                      `json:"bindAsFiles"`
+	NamingStrategy         string                    `json:"namingStrategy,omitempty"`
+}
+
+type ServiceBindingReference struct {
+	Kind       string `json:"kind,omitempty"`
+	Resource   string `json:"resource,omitempty"`
+	Namespace  string `json:"namespace,omitempty"`
+	Name       string `json:"name,omitempty"`
+	APIVersion string `json:"apiVersion,omitempty"`
 }
 
 type ServiceBindingStatus struct {

--- a/pkg/binding/binding.go
+++ b/pkg/binding/binding.go
@@ -29,6 +29,7 @@ import (
 	backendpkg "github.com/redhat-developer/odo/pkg/binding/backend"
 	"github.com/redhat-developer/odo/pkg/kclient"
 	"github.com/redhat-developer/odo/pkg/libdevfile"
+	clierrors "github.com/redhat-developer/odo/pkg/odo/cli/errors"
 )
 
 type BindingClient struct {
@@ -114,6 +115,7 @@ func (o *BindingClient) GetServiceInstances(namespace string) (map[string]unstru
 
 // GetBindingsFromDevfile returns all ServiceBinding resources declared as Kubernertes component from a Devfile
 // from group binding.operators.coreos.com/v1alpha1 or servicebinding.io/v1alpha3
+// The function also returns status information of the binding in the cluster, if accessible, or a warning if the cluster is not accessible
 func (o *BindingClient) GetBindingsFromDevfile(devfileObj parser.DevfileObj, context string) ([]api.ServiceBinding, error) {
 	result := []api.ServiceBinding{}
 	kubeComponents, err := devfileObj.Data.GetComponents(parsercommon.DevfileOptions{
@@ -125,6 +127,7 @@ func (o *BindingClient) GetBindingsFromDevfile(devfileObj parser.DevfileObj, con
 		return nil, err
 	}
 
+	var warning error
 	for _, component := range kubeComponents {
 		strCRD, err := libdevfile.GetK8sManifestsWithVariablesSubstituted(devfileObj, component.Name, context, devfilefs.DefaultFs{})
 		if err != nil {
@@ -145,13 +148,13 @@ func (o *BindingClient) GetBindingsFromDevfile(devfileObj parser.DevfileObj, con
 				return nil, err
 			}
 
-			sb, err := o.kubernetesClient.APIServiceBindingFromBinding(sbo)
+			sb, err := kclient.APIServiceBindingFromBinding(sbo)
 			if err != nil {
 				return nil, err
 			}
 			sb.Status, err = o.getStatusFromBinding(sb.Name)
 			if err != nil {
-				return nil, err
+				warning = clierrors.NewWarning("unable to access the cluster", err)
 			}
 
 			result = append(result, sb)
@@ -164,27 +167,26 @@ func (o *BindingClient) GetBindingsFromDevfile(devfileObj parser.DevfileObj, con
 				return nil, err
 			}
 
-			sb := o.kubernetesClient.APIServiceBindingFromSpec(sbc)
+			sb := kclient.APIServiceBindingFromSpec(sbc)
 			sb.Status, err = o.getStatusFromSpec(sb.Name)
 			if err != nil {
-				return nil, err
+				warning = clierrors.NewWarning("unable to access the cluster", err)
 			}
 
 			result = append(result, sb)
 
 		}
 	}
-	return result, nil
+	return result, warning
 }
 
 // GetBindingFromCluster returns the ServiceBinding resource with the given name
 // from the cluster, from group binding.operators.coreos.com/v1alpha1 or servicebinding.io/v1alpha3
 func (o *BindingClient) GetBindingFromCluster(name string) (api.ServiceBinding, error) {
-
 	bindingSB, err := o.kubernetesClient.GetBindingServiceBinding(name)
 	if err == nil {
 		var sb api.ServiceBinding
-		sb, err = o.kubernetesClient.APIServiceBindingFromBinding(bindingSB)
+		sb, err = kclient.APIServiceBindingFromBinding(bindingSB)
 		if err != nil {
 			return api.ServiceBinding{}, err
 		}
@@ -200,7 +202,7 @@ func (o *BindingClient) GetBindingFromCluster(name string) (api.ServiceBinding, 
 
 	specSB, err := o.kubernetesClient.GetSpecServiceBinding(name)
 	if err == nil {
-		sb := o.kubernetesClient.APIServiceBindingFromSpec(specSB)
+		sb := kclient.APIServiceBindingFromSpec(specSB)
 		sb.Status, err = o.getStatusFromSpec(specSB.Name)
 		if err != nil {
 			return api.ServiceBinding{}, err
@@ -218,6 +220,9 @@ func (o *BindingClient) GetBindingFromCluster(name string) (api.ServiceBinding, 
 // getStatusFromBinding returns status information from a ServiceBinding in the cluster
 // from group binding.operators.coreos.com/v1alpha1
 func (o *BindingClient) getStatusFromBinding(name string) (*api.ServiceBindingStatus, error) {
+	if o.kubernetesClient == nil {
+		return nil, nil
+	}
 	bindingSB, err := o.kubernetesClient.GetBindingServiceBinding(name)
 	if err != nil {
 		if kerrors.IsNotFound(err) {

--- a/pkg/binding/list_test.go
+++ b/pkg/binding/list_test.go
@@ -21,12 +21,12 @@ import (
 var apiServiceBinding = api.ServiceBinding{
 	Name: "my-nodejs-app-cluster-sample",
 	Spec: api.ServiceBindingSpec{
-		Application: corev1.ObjectReference{
+		Application: api.ServiceBindingReference{
 			Name:       "my-nodejs-app-app",
 			APIVersion: deploymentApiVersion,
 			Kind:       deploymentKind,
 		},
-		Services: []corev1.ObjectReference{
+		Services: []api.ServiceBindingReference{
 			{
 				Name:       "cluster-sample",
 				APIVersion: ClusterAPIVersion,
@@ -109,7 +109,6 @@ func TestBindingClient_ListAllBindings(t *testing.T) {
 							"my-nodejs-app-cluster-sample",
 						),
 					)
-					client.EXPECT().APIServiceBindingFromBinding(gomock.Any()).Return(apiServiceBinding, nil)
 					return client
 				},
 			}, args: args{
@@ -133,7 +132,6 @@ func TestBindingClient_ListAllBindings(t *testing.T) {
 					).AnyTimes()
 					client.EXPECT().GetCurrentNamespace().Return("anamespace").AnyTimes()
 					client.EXPECT().GetSecret("asecret", "anamespace").Return(&sbSecret, nil).AnyTimes()
-					client.EXPECT().APIServiceBindingFromBinding(gomock.Any()).Return(apiServiceBinding, nil).AnyTimes()
 					return client
 				},
 			}, args: args{
@@ -144,12 +142,12 @@ func TestBindingClient_ListAllBindings(t *testing.T) {
 				{
 					Name: "my-nodejs-app-cluster-sample",
 					Spec: api.ServiceBindingSpec{
-						Application: corev1.ObjectReference{
+						Application: api.ServiceBindingReference{
 							Name:       "my-nodejs-app-app",
 							APIVersion: deploymentApiVersion,
 							Kind:       deploymentKind,
 						},
-						Services: []corev1.ObjectReference{
+						Services: []api.ServiceBindingReference{
 							{
 								Name:       "cluster-sample",
 								APIVersion: ClusterAPIVersion,
@@ -181,7 +179,6 @@ func TestBindingClient_ListAllBindings(t *testing.T) {
 					).AnyTimes()
 					client.EXPECT().GetCurrentNamespace().Return("anamespace").AnyTimes()
 					client.EXPECT().GetSecret("asecret", "anamespace").Return(&sbSecret, nil).AnyTimes()
-					client.EXPECT().APIServiceBindingFromBinding(gomock.Any()).Return(apiServiceBinding, nil).AnyTimes()
 					return client
 				},
 			}, args: args{},
@@ -189,12 +186,12 @@ func TestBindingClient_ListAllBindings(t *testing.T) {
 				{
 					Name: "my-nodejs-app-cluster-sample",
 					Spec: api.ServiceBindingSpec{
-						Application: corev1.ObjectReference{
+						Application: api.ServiceBindingReference{
 							Name:       "my-nodejs-app-app",
 							APIVersion: deploymentApiVersion,
 							Kind:       deploymentKind,
 						},
-						Services: []corev1.ObjectReference{
+						Services: []api.ServiceBindingReference{
 							{
 								Name:       "cluster-sample",
 								APIVersion: ClusterAPIVersion,

--- a/pkg/init/asker/mock.go
+++ b/pkg/init/asker/mock.go
@@ -157,22 +157,6 @@ func (mr *MockAskerMockRecorder) AskStarterProject(projects interface{}) *gomock
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "AskStarterProject", reflect.TypeOf((*MockAsker)(nil).AskStarterProject), projects)
 }
 
-// AskVersion mocks base method
-func (m *MockAsker) AskVersion(versions []api.DevfileStackVersion) (bool, string, error) {
-	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "AskVersion", versions)
-	ret0, _ := ret[0].(bool)
-	ret1, _ := ret[1].(string)
-	ret2, _ := ret[2].(error)
-	return ret0, ret1, ret2
-}
-
-// AskVersion indicates an expected call of AskVersion
-func (mr *MockAskerMockRecorder) AskVersion(versions interface{}) *gomock.Call {
-	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "AskVersion", reflect.TypeOf((*MockAsker)(nil).AskVersion), versions)
-}
-
 // AskType mocks base method.
 func (m *MockAsker) AskType(types registry.TypesWithDetails) (bool, api.DevfileStack, error) {
 	m.ctrl.T.Helper()
@@ -187,4 +171,20 @@ func (m *MockAsker) AskType(types registry.TypesWithDetails) (bool, api.DevfileS
 func (mr *MockAskerMockRecorder) AskType(types interface{}) *gomock.Call {
 	mr.mock.ctrl.T.Helper()
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "AskType", reflect.TypeOf((*MockAsker)(nil).AskType), types)
+}
+
+// AskVersion mocks base method.
+func (m *MockAsker) AskVersion(versions []api.DevfileStackVersion) (bool, string, error) {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "AskVersion", versions)
+	ret0, _ := ret[0].(bool)
+	ret1, _ := ret[1].(string)
+	ret2, _ := ret[2].(error)
+	return ret0, ret1, ret2
+}
+
+// AskVersion indicates an expected call of AskVersion.
+func (mr *MockAskerMockRecorder) AskVersion(versions interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "AskVersion", reflect.TypeOf((*MockAsker)(nil).AskVersion), versions)
 }

--- a/pkg/kclient/binding.go
+++ b/pkg/kclient/binding.go
@@ -238,11 +238,13 @@ func (c Client) ListServiceBindingsFromAllGroups() ([]specApi.ServiceBinding, []
 
 // APIServiceBindingFromBinding returns a common api.ServiceBinding structure
 // from a ServiceBinding.binding.operators.coreos.com/v1alpha1
-func (c Client) APIServiceBindingFromBinding(binding bindingApi.ServiceBinding) (api.ServiceBinding, error) {
+func APIServiceBindingFromBinding(
+	binding bindingApi.ServiceBinding,
+) (api.ServiceBinding, error) {
 
-	var dstSvcs []corev1.ObjectReference
+	var dstSvcs []api.ServiceBindingReference
 	for _, srcSvc := range binding.Spec.Services {
-		dstSvc := corev1.ObjectReference{
+		dstSvc := api.ServiceBindingReference{
 			Name: srcSvc.Name,
 		}
 		dstSvc.APIVersion, dstSvc.Kind = schema.GroupVersion{
@@ -256,21 +258,11 @@ func (c Client) APIServiceBindingFromBinding(binding bindingApi.ServiceBinding) 
 	}
 
 	application := binding.Spec.Application
-	refToApplication := corev1.ObjectReference{
-		Name: application.Name,
+	refToApplication := api.ServiceBindingReference{
+		Name:     application.Name,
+		Resource: application.Resource,
 	}
 
-	if application.Kind == "" {
-		gvk, err := c.GetGVKFromGVR(schema.GroupVersionResource{
-			Group:    application.Group,
-			Version:  application.Version,
-			Resource: application.Resource,
-		})
-		if err != nil {
-			return api.ServiceBinding{}, err
-		}
-		application.Kind = gvk.Kind
-	}
 	refToApplication.APIVersion, refToApplication.Kind = schema.GroupVersion{
 		Group:   application.Group,
 		Version: application.Version,
@@ -290,17 +282,17 @@ func (c Client) APIServiceBindingFromBinding(binding bindingApi.ServiceBinding) 
 
 // APIServiceBindingFromSpec returns a common api.ServiceBinding structure
 // from a ServiceBinding.servicebinding.io/v1alpha3
-func (c Client) APIServiceBindingFromSpec(spec specApi.ServiceBinding) api.ServiceBinding {
+func APIServiceBindingFromSpec(spec specApi.ServiceBinding) api.ServiceBinding {
 
 	service := spec.Spec.Service
-	refToService := corev1.ObjectReference{
+	refToService := api.ServiceBindingReference{
 		APIVersion: service.APIVersion,
 		Kind:       service.Kind,
 		Name:       service.Name,
 	}
 
 	application := spec.Spec.Workload
-	refToApplication := corev1.ObjectReference{
+	refToApplication := api.ServiceBindingReference{
 		APIVersion: application.APIVersion,
 		Kind:       application.Kind,
 		Name:       application.Name,
@@ -310,7 +302,7 @@ func (c Client) APIServiceBindingFromSpec(spec specApi.ServiceBinding) api.Servi
 		Name: spec.Name,
 		Spec: api.ServiceBindingSpec{
 			Application:            refToApplication,
-			Services:               []corev1.ObjectReference{refToService},
+			Services:               []api.ServiceBindingReference{refToService},
 			DetectBindingResources: false,
 			BindAsFiles:            true,
 		},

--- a/pkg/kclient/interface.go
+++ b/pkg/kclient/interface.go
@@ -24,8 +24,6 @@ import (
 
 	bindingApi "github.com/redhat-developer/service-binding-operator/apis/binding/v1alpha1"
 	specApi "github.com/redhat-developer/service-binding-operator/apis/spec/v1alpha3"
-
-	"github.com/redhat-developer/odo/pkg/api"
 )
 
 type ClientInterface interface {
@@ -43,8 +41,6 @@ type ClientInterface interface {
 	GetSpecServiceBinding(name string) (specApi.ServiceBinding, error)
 	ListServiceBindingsFromAllGroups() ([]specApi.ServiceBinding, []bindingApi.ServiceBinding, error)
 	NewServiceBindingServiceObject(serviceNs string, unstructuredService unstructured.Unstructured, bindingName string) (bindingApi.Service, error)
-	APIServiceBindingFromBinding(binding bindingApi.ServiceBinding) (api.ServiceBinding, error)
-	APIServiceBindingFromSpec(spec specApi.ServiceBinding) api.ServiceBinding
 	GetWorkloadKinds() ([]string, []schema.GroupVersionKind, error)
 
 	// deployment.go

--- a/pkg/kclient/mock_Client.go
+++ b/pkg/kclient/mock_Client.go
@@ -14,7 +14,6 @@ import (
 	gomock "github.com/golang/mock/gomock"
 	v1 "github.com/openshift/api/project/v1"
 	v1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
-	api "github.com/redhat-developer/odo/pkg/api"
 	v1alpha10 "github.com/redhat-developer/service-binding-operator/apis/binding/v1alpha1"
 	v1alpha3 "github.com/redhat-developer/service-binding-operator/apis/spec/v1alpha3"
 	v10 "k8s.io/api/apps/v1"
@@ -53,35 +52,6 @@ func NewMockClientInterface(ctrl *gomock.Controller) *MockClientInterface {
 // EXPECT returns an object that allows the caller to indicate expected use.
 func (m *MockClientInterface) EXPECT() *MockClientInterfaceMockRecorder {
 	return m.recorder
-}
-
-// APIServiceBindingFromBinding mocks base method.
-func (m *MockClientInterface) APIServiceBindingFromBinding(binding v1alpha10.ServiceBinding) (api.ServiceBinding, error) {
-	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "APIServiceBindingFromBinding", binding)
-	ret0, _ := ret[0].(api.ServiceBinding)
-	ret1, _ := ret[1].(error)
-	return ret0, ret1
-}
-
-// APIServiceBindingFromBinding indicates an expected call of APIServiceBindingFromBinding.
-func (mr *MockClientInterfaceMockRecorder) APIServiceBindingFromBinding(binding interface{}) *gomock.Call {
-	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "APIServiceBindingFromBinding", reflect.TypeOf((*MockClientInterface)(nil).APIServiceBindingFromBinding), binding)
-}
-
-// APIServiceBindingFromSpec mocks base method.
-func (m *MockClientInterface) APIServiceBindingFromSpec(spec v1alpha3.ServiceBinding) api.ServiceBinding {
-	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "APIServiceBindingFromSpec", spec)
-	ret0, _ := ret[0].(api.ServiceBinding)
-	return ret0
-}
-
-// APIServiceBindingFromSpec indicates an expected call of APIServiceBindingFromSpec.
-func (mr *MockClientInterfaceMockRecorder) APIServiceBindingFromSpec(spec interface{}) *gomock.Call {
-	mr.mock.ctrl.T.Helper()
-	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "APIServiceBindingFromSpec", reflect.TypeOf((*MockClientInterface)(nil).APIServiceBindingFromSpec), spec)
 }
 
 // ApplyDeployment mocks base method.

--- a/pkg/odo/cli/describe/binding.go
+++ b/pkg/odo/cli/describe/binding.go
@@ -2,6 +2,7 @@ package describe
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strconv"
 
@@ -11,6 +12,7 @@ import (
 
 	"github.com/redhat-developer/odo/pkg/api"
 	"github.com/redhat-developer/odo/pkg/log"
+	clierrors "github.com/redhat-developer/odo/pkg/odo/cli/errors"
 	"github.com/redhat-developer/odo/pkg/odo/cmdline"
 	"github.com/redhat-developer/odo/pkg/odo/commonflags"
 	odocontext "github.com/redhat-developer/odo/pkg/odo/context"
@@ -96,11 +98,30 @@ func (o *BindingOptions) runWithoutName(ctx context.Context) ([]api.ServiceBindi
 		devfileObj = odocontext.GetDevfileObj(ctx)
 	)
 
-	return o.clientset.BindingClient.GetBindingsFromDevfile(*devfileObj, workingDir)
+	result, err := o.clientset.BindingClient.GetBindingsFromDevfile(*devfileObj, workingDir)
+	if err != nil {
+		if clierrors.AsWarning(err) {
+			log.Warning(err.Error())
+			return result, nil
+		}
+		return nil, err
+	}
+	return result, nil
 }
 
 func (o *BindingOptions) runWithName() (api.ServiceBinding, error) {
-	return o.clientset.BindingClient.GetBindingFromCluster(o.nameFlag)
+	if o.clientset.KubernetesClient == nil {
+		return api.ServiceBinding{}, errors.New("unable to access the cluster")
+	}
+	result, err := o.clientset.BindingClient.GetBindingFromCluster(o.nameFlag)
+	if err != nil {
+		if clierrors.AsWarning(err) {
+			log.Warning(err.Error())
+			return result, nil
+		}
+		return api.ServiceBinding{}, err
+	}
+	return result, nil
 }
 
 // NewCmdBinding implements the binding odo sub-command

--- a/pkg/odo/cli/list/binding/binding.go
+++ b/pkg/odo/cli/list/binding/binding.go
@@ -55,7 +55,7 @@ func (o *BindingListOptions) SetClientset(clientset *clientset.Clientset) {
 
 // Complete completes BindingListOptions after they've been created
 func (o *BindingListOptions) Complete(ctx context.Context, cmdline cmdline.Cmdline, args []string) (err error) {
-	if o.namespaceFlag != "" {
+	if o.clientset.KubernetesClient != nil && o.namespaceFlag != "" {
 		o.clientset.KubernetesClient.SetNamespace(o.namespaceFlag)
 	}
 	return nil

--- a/pkg/odo/cli/list/binding/binding.go
+++ b/pkg/odo/cli/list/binding/binding.go
@@ -68,7 +68,7 @@ func (o *BindingListOptions) Validate(ctx context.Context) (err error) {
 
 // Run contains the logic for the odo list binding command
 func (o *BindingListOptions) Run(ctx context.Context) error {
-	listSpinner := log.Spinnerf("Listing ServiceBindings from the namespace %q", o.clientset.KubernetesClient.GetCurrentNamespace())
+	listSpinner := log.Spinnerf("Listing ServiceBindings")
 	defer listSpinner.End(false)
 
 	list, err := o.run(ctx)
@@ -78,7 +78,7 @@ func (o *BindingListOptions) Run(ctx context.Context) error {
 
 	listSpinner.End(true)
 
-	HumanReadableOutput(o.clientset.KubernetesClient.GetCurrentNamespace(), list)
+	HumanReadableOutput(list)
 	return nil
 }
 
@@ -123,10 +123,10 @@ func NewCmdBindingList(name, fullName string) *cobra.Command {
 }
 
 // HumanReadableOutput outputs the list of bindings in a human readable format
-func HumanReadableOutput(namespace string, list api.ResourcesList) {
+func HumanReadableOutput(list api.ResourcesList) {
 	bindings := list.Bindings
 	if len(bindings) == 0 {
-		log.Errorf("There are no service bindings in the %q namespace.", namespace)
+		log.Errorf("There are no service bindings.")
 		return
 
 	}

--- a/pkg/odo/cli/list/binding/binding.go
+++ b/pkg/odo/cli/list/binding/binding.go
@@ -152,7 +152,12 @@ func HumanReadableOutput(namespace string, list api.ResourcesList) {
 		}
 
 		appSpec := binding.Spec.Application
-		application := fmt.Sprintf("%s (%s)", appSpec.Name, appSpec.Kind)
+		var application string
+		if appSpec.Kind != "" {
+			application = fmt.Sprintf("%s (%s)", appSpec.Name, appSpec.Kind)
+		} else {
+			application = fmt.Sprintf("%s (%s)", appSpec.Name, appSpec.Resource)
+		}
 
 		servicesSpecs := binding.Spec.Services
 		services := ""

--- a/pkg/odo/cli/list/list.go
+++ b/pkg/odo/cli/list/list.go
@@ -102,7 +102,7 @@ func (lo *ListOptions) Run(ctx context.Context) error {
 	fmt.Printf("\nComponents:\n")
 	clicomponent.HumanReadableOutput(ctx, list)
 	fmt.Printf("\nBindings:\n")
-	binding.HumanReadableOutput(lo.namespaceFilter, list)
+	binding.HumanReadableOutput(list)
 	return nil
 }
 

--- a/tests/examples/source/devfiles/nodejs/devfile-with-service-binding-envvars.yaml
+++ b/tests/examples/source/devfiles/nodejs/devfile-with-service-binding-envvars.yaml
@@ -56,7 +56,7 @@ components:
         application:
           group: apps
           name: my-nodejs-app-app
-          resource: deployments
+          kind: Deployment
           version: v1
         bindAsFiles: false
         detectBindingResources: true

--- a/tests/integration/cmd_describe_list_binding_test.go
+++ b/tests/integration/cmd_describe_list_binding_test.go
@@ -126,7 +126,7 @@ var _ = Describe("odo describe/list binding command tests", func() {
 							Expect(lines[0]).To(ContainSubstring(fmt.Sprintf("Listing resources from the namespace %q", commonVar.Project)))
 							lines = lines[6:]
 						} else {
-							Expect(lines[0]).To(ContainSubstring(fmt.Sprintf("Listing ServiceBindings from the namespace %q", commonVar.Project)))
+							Expect(lines[0]).To(ContainSubstring("Listing ServiceBindings"))
 						}
 						Expect(lines[3]).To(ContainSubstring("* "))
 						Expect(lines[3]).To(ContainSubstring("my-nodejs-app-cluster-sample"))
@@ -227,7 +227,7 @@ var _ = Describe("odo describe/list binding command tests", func() {
 				},
 				assertListHumanReadableOutput: func(devfile bool, stdout, stderr string) {
 					lines := strings.Split(stdout, "\n")
-					Expect(lines[0]).To(ContainSubstring(fmt.Sprintf("Listing ServiceBindings from the namespace %q", commonVar.Project)))
+					Expect(lines[0]).To(ContainSubstring("Listing ServiceBindings"))
 					if devfile {
 						Expect(lines[3]).To(ContainSubstring("* "))
 					} else {
@@ -318,7 +318,7 @@ var _ = Describe("odo describe/list binding command tests", func() {
 				},
 				assertListHumanReadableOutput: func(devfile bool, stdout, stderr string) {
 					lines := strings.Split(stdout, "\n")
-					Expect(lines[0]).To(ContainSubstring(fmt.Sprintf("Listing ServiceBindings from the namespace %q", commonVar.Project)))
+					Expect(lines[0]).To(ContainSubstring("Listing ServiceBindings"))
 					if devfile {
 						Expect(lines[3]).To(ContainSubstring("* "))
 					} else {
@@ -397,7 +397,7 @@ var _ = Describe("odo describe/list binding command tests", func() {
 				},
 				assertListHumanReadableOutput: func(devfile bool, stdout, stderr string) {
 					lines := strings.Split(stdout, "\n")
-					Expect(lines[0]).To(ContainSubstring(fmt.Sprintf("Listing ServiceBindings from the namespace %q", commonVar.Project)))
+					Expect(lines[0]).To(ContainSubstring("Listing ServiceBindings"))
 					if devfile {
 						Expect(lines[3]).To(ContainSubstring("* "))
 					} else {
@@ -473,7 +473,7 @@ var _ = Describe("odo describe/list binding command tests", func() {
 				},
 				assertListHumanReadableOutput: func(devfile bool, stdout, stderr string) {
 					lines := strings.Split(stdout, "\n")
-					Expect(lines[0]).To(ContainSubstring(fmt.Sprintf("Listing ServiceBindings from the namespace %q", commonVar.Project)))
+					Expect(lines[0]).To(ContainSubstring("Listing ServiceBindings"))
 					if devfile {
 						Expect(lines[3]).To(ContainSubstring("* "))
 					} else {

--- a/tests/integration/cmd_describe_list_binding_test.go
+++ b/tests/integration/cmd_describe_list_binding_test.go
@@ -41,48 +41,52 @@ var _ = Describe("odo describe/list binding command tests", func() {
 				}
 			})
 
-			It("should describe the binding without running odo dev", func() {
-				By("JSON output", func() {
-					res := helper.Cmd("odo", "describe", "binding", "-o", "json").ShouldPass()
-					stdout, stderr := res.Out(), res.Err()
-					Expect(stderr).To(BeEmpty())
-					Expect(helper.IsJSON(stdout)).To(BeTrue())
-					helper.JsonPathContentIs(stdout, "0.name", "my-nodejs-app-cluster-sample")
-					helper.JsonPathContentIs(stdout, "0.spec.application.kind", "Deployment")
-					helper.JsonPathContentIs(stdout, "0.spec.application.name", "my-nodejs-app-app")
-					helper.JsonPathContentIs(stdout, "0.spec.application.apiVersion", "apps/v1")
-					helper.JsonPathContentIs(stdout, "0.spec.services.0.apiVersion", "postgresql.k8s.enterprisedb.io/v1")
-					helper.JsonPathContentIs(stdout, "0.spec.services.0.kind", "Cluster")
-					helper.JsonPathContentIs(stdout, "0.spec.services.0.name", "cluster-sample")
-					if ns != "" {
-						helper.JsonPathContentIs(stdout, "0.spec.services.0.namespace", ns)
-					} else {
-						helper.JsonPathDoesNotExist(stdout, "0.spec.services.0.namespace")
-					}
-					helper.JsonPathContentIs(stdout, "0.spec.detectBindingResources", "true")
-					helper.JsonPathContentIs(stdout, "0.spec.bindAsFiles", "true")
-					helper.JsonPathContentIs(stdout, "0.spec.namingStrategy", "lowercase")
-					helper.JsonPathContentIs(stdout, "0.status", "")
+			for _, label := range []string{
+				helper.LabelNoCluster, helper.LabelUnauth,
+			} {
+				label := label
+				It("should describe the binding without running odo dev", Label(label), func() {
+					By("JSON output", func() {
+						res := helper.Cmd("odo", "describe", "binding", "-o", "json").ShouldPass()
+						stdout, stderr := res.Out(), res.Err()
+						Expect(stderr).To(BeEmpty())
+						Expect(helper.IsJSON(stdout)).To(BeTrue())
+						helper.JsonPathContentIs(stdout, "0.name", "my-nodejs-app-cluster-sample")
+						helper.JsonPathContentIs(stdout, "0.spec.application.kind", "Deployment")
+						helper.JsonPathContentIs(stdout, "0.spec.application.name", "my-nodejs-app-app")
+						helper.JsonPathContentIs(stdout, "0.spec.application.apiVersion", "apps/v1")
+						helper.JsonPathContentIs(stdout, "0.spec.services.0.apiVersion", "postgresql.k8s.enterprisedb.io/v1")
+						helper.JsonPathContentIs(stdout, "0.spec.services.0.kind", "Cluster")
+						helper.JsonPathContentIs(stdout, "0.spec.services.0.name", "cluster-sample")
+						if ns != "" {
+							helper.JsonPathContentIs(stdout, "0.spec.services.0.namespace", ns)
+						} else {
+							helper.JsonPathDoesNotExist(stdout, "0.spec.services.0.namespace")
+						}
+						helper.JsonPathContentIs(stdout, "0.spec.detectBindingResources", "true")
+						helper.JsonPathContentIs(stdout, "0.spec.bindAsFiles", "true")
+						helper.JsonPathContentIs(stdout, "0.spec.namingStrategy", "lowercase")
+						helper.JsonPathContentIs(stdout, "0.status", "")
+					})
+					By("human readable output", func() {
+						res := helper.Cmd("odo", "describe", "binding").ShouldPass()
+						stdout, _ := res.Out(), res.Err()
+						Expect(stdout).To(ContainSubstring("ServiceBinding used by the current component"))
+						Expect(stdout).To(ContainSubstring("Service Binding Name: my-nodejs-app-cluster-sample"))
+						if ns != "" {
+							Expect(stdout).To(ContainSubstring(fmt.Sprintf("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io) (namespace: %s)", ns)))
+						} else {
+							Expect(stdout).To(ContainSubstring("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io)"))
+							Expect(stdout).ToNot(ContainSubstring("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io) (namespace: "))
+						}
+						Expect(stdout).To(ContainSubstring("Bind as files: true"))
+						Expect(stdout).To(ContainSubstring("Detect binding resources: true"))
+						Expect(stdout).To(ContainSubstring("Naming strategy: lowercase"))
+						Expect(stdout).To(ContainSubstring("Available binding information: unknown"))
+						Expect(stdout).To(ContainSubstring("Binding information for one or more ServiceBinding is not available"))
+					})
 				})
-				By("human readable output", func() {
-					res := helper.Cmd("odo", "describe", "binding").ShouldPass()
-					stdout, _ := res.Out(), res.Err()
-					Expect(stdout).To(ContainSubstring("ServiceBinding used by the current component"))
-					Expect(stdout).To(ContainSubstring("Service Binding Name: my-nodejs-app-cluster-sample"))
-					if ns != "" {
-						Expect(stdout).To(ContainSubstring(fmt.Sprintf("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io) (namespace: %s)", ns)))
-					} else {
-						Expect(stdout).To(ContainSubstring("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io)"))
-						Expect(stdout).ToNot(ContainSubstring("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io) (namespace: "))
-					}
-					Expect(stdout).To(ContainSubstring("Bind as files: true"))
-					Expect(stdout).To(ContainSubstring("Detect binding resources: true"))
-					Expect(stdout).To(ContainSubstring("Naming strategy: lowercase"))
-					Expect(stdout).To(ContainSubstring("Available binding information: unknown"))
-					Expect(stdout).To(ContainSubstring("Binding information for one or more ServiceBinding is not available"))
-				})
-			})
-
+			}
 			for _, command := range [][]string{
 				{"list", "binding"},
 				{"list"},

